gossip/src/cluster_info.rs

```diff
@@ -2440,32 +2440,87 @@ impl ClusterInfo {
-    // verifying them and then sending them down the channel for the actual
-    // handling of requests/messages.
-    fn run_socket_consume(
+    fn handle_packet_batch_receive(
         &self,
         receiver: &PacketBatchReceiver,
-        sender: &Sender<Vec<(/*from:*/ SocketAddr, Protocol)>>,
-        thread_pool: &ThreadPool,
-    ) -> Result<(), GossipError> {
+        max_allowed_packets: usize,
+    ) -> Result<(VecDeque<PacketBatch>, usize), GossipError> {
         const RECV_TIMEOUT: Duration = Duration::from_secs(1);
-        let packets: Vec<_> = receiver.recv_timeout(RECV_TIMEOUT)?.packets.into();
-        let mut packets = VecDeque::from(packets);
+        let packet_batch = receiver.recv_timeout(RECV_TIMEOUT)?;
+        let mut total_num_packets_received = packet_batch.packets.len();
+        let mut first_valid_packet_in_oldest_batch = 0;
+        let mut packet_batches = VecDeque::new();
+        packet_batches.push_back(packet_batch);
+
         for packet_batch in receiver.try_iter() {
-            packets.extend(packet_batch.packets.iter().cloned());
-            let excess_count = packets.len().saturating_sub(MAX_GOSSIP_TRAFFIC);
+            total_num_packets_received =
+                total_num_packets_received.saturating_add(packet_batch.packets.len());
+            packet_batches.push_back(packet_batch);
+
+            let mut excess_count = total_num_packets_received.saturating_sub(max_allowed_packets);
             if excess_count > 0 {
-                packets.drain(0..excess_count);
                 self.stats
                     .gossip_packets_dropped_count
                     .add_relaxed(excess_count as u64);
+
+                while excess_count > 0 {
+                    // There must be at least one batch present if there is excess
+                    let oldest_batch_num_packets_remaining = packet_batches
+                        .front()
+                        .unwrap()
+                        .packets
+                        .len()
+                        .saturating_sub(first_valid_packet_in_oldest_batch);
+                    if oldest_batch_num_packets_remaining > excess_count {
+                        // There are enough packets in oldest batch to drop to get
+                        // under the threshold for max packets
+                        first_valid_packet_in_oldest_batch =
+                            first_valid_packet_in_oldest_batch.saturating_add(excess_count);
+                        break;
+                    } else {
+                        // This batch has been completely drained, drop it and
+                        // start removing excess from next oldest batch
+                        excess_count =
+                            excess_count.saturating_sub(oldest_batch_num_packets_remaining);
+                        first_valid_packet_in_oldest_batch = 0;
+                        packet_batches.pop_front();
+                    }
```

Comment:
I think this can be simplified by something like:
```rust
let front_size = packet_batches.front().unwrap().packets.len();
let discard_count = usize::min(excess_count, 
    front_size - first_valid_packet_in_oldest_batch);
excess_count -= discard_count;
first_valid_packet_in_oldest_batch += discard_count;
if first_valid_packet_in_oldest_batch >= front_size {
   packet_batches.pop_front();
   first_valid_packet_in_oldest_batch = 0;
}
```